bucket_map/src/bucket_storage.rs

```diff
@@ -337,57 +337,79 @@ impl<O: BucketOccupied> BucketStorage<O> {
-
-        // Theoretical performance optimization: write a zero to the end of
-        // the file so that we won't have to resize it later, which may be
-        // expensive.
-        //debug!("GROWING file {}", capacity * cell_size as u64);
-        data.seek(SeekFrom::Start(bytes - 1)).unwrap();
-        data.write_all(&[0]).unwrap();
-        data.rewind().unwrap();
-        measure_new_file.stop();
-        let mut measure_flush = Measure::start("measure_flush");
-        data.flush().unwrap(); // can we skip this?
-        measure_flush.stop();
+            .create(create)
+            .open(path.clone());
+        if let Err(e) = data {
+            if !create {
+                // we can't load this file, so bail without error
+                return None;
+            }
+            panic!(
+                "Unable to create data file {:?} in current dir({:?}): {:?}",
+                path,
+                std::env::current_dir(),
+                e
+            );
+        }
+        let mut data = data.unwrap();
+
+        if create {
+            // Theoretical performance optimization: write a zero to the end of
+            // the file so that we won't have to resize it later, which may be
+            // expensive.
+            //debug!("GROWING file {}", capacity * cell_size as u64);
+            data.seek(SeekFrom::Start(create_bytes - 1)).unwrap();
+            data.write_all(&[0]).unwrap();
+            data.rewind().unwrap();
+            measure_new_file.stop();
+            let measure_flush = Measure::start("measure_flush");
+            data.flush().unwrap(); // can we skip this?
+            stats
+                .flush_file_us
+                .fetch_add(measure_flush.end_as_us(), Ordering::Relaxed);
+        }
         let mut measure_mmap = Measure::start("measure_mmap");
-        let res = (
-            unsafe { MmapMut::map_mut(&data).unwrap() },
-            file,
-            file_random,
-        );
+        let res = unsafe { MmapMut::map_mut(&data) };
+        if let Err(e) = res {
+            panic!(
+                "Unable to mmap file {:?} in current dir({:?}): {:?}",
+                path,
+                std::env::current_dir(),
+                e
+            );
+        }
         measure_mmap.stop();
         stats
             .new_file_us
             .fetch_add(measure_new_file.as_us(), Ordering::Relaxed);
-        stats
-            .flush_file_us
-            .fetch_add(measure_flush.as_us(), Ordering::Relaxed);
         stats
             .mmap_us
             .fetch_add(measure_mmap.as_us(), Ordering::Relaxed);
-        res
+        res.ok()
+    }
+
+    /// allocate a new memory mapped file of size `bytes` on one of `drives`
+    fn new_map(drives: &[PathBuf], bytes: u64, stats: &BucketStats) -> (MmapMut, PathBuf, u128) {
+        let r = thread_rng().gen_range(0..drives.len());
+        let drive = &drives[r];
+        let file_random = thread_rng().gen_range(0..u128::MAX);
```

Comment:
If `file_random` produces the same number in two calls, `map_open_file` will panic.
This is very unlikely. Maybe we don't need to handle this case. Anyway just a note.